fix(rocprim): arrow operator iterators

NB4444 opened the pull request:
## Motivation

The '->' operator currently does not compile when used for `transform_iterator`, the `texture_cache_iterator`, the `zip_iterator` and the `arg_index_iterator`.

## Technical Details

Fixed the `->` operator for the `transform_iterator`, the `texture_cache_iterator` and the `arg_index_iterator`, by now returning a proxy pointer. It did not compile before when used.

Deprecated the  `->` operator for the `zip_iterator`. It did not work and there is no useful implementation for this operator, CUB also does not have this operator.

## Test Plan

Extra tests have been added for  the `transform_iterator`, the `texture_cache_iterator` and the `arg_index_iterator` with this operator.

## Test Result

They all pass.

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

amd-hsong commented on projects/rocprim/rocprim/include/rocprim/iterator/arg_index_iterator.hpp:
```diff
@@ -58,6 +58,29 @@ class arg_index_iterator
+        T value_;
+
+    public:
+        ROCPRIM_HOST_DEVICE inline proxy_pointer(const T& value) : value_(value) {}
```
Consider adding an "explicit" keyword here

amd-hsong commented on projects/rocprim/rocprim/include/rocprim/iterator/texture_cache_iterator.hpp:
```diff
@@ -132,13 +132,37 @@ template<
+        V value_;
+
+    public:
+        ROCPRIM_HOST_DEVICE inline proxy_pointer(const V& value) : value_(value) {}
```
Consider adding an "explicit" keyword here

amd-hsong commented on projects/rocprim/rocprim/include/rocprim/iterator/transform_iterator.hpp:
```diff
@@ -54,22 +54,47 @@ template<class InputIterator,
+        T value_;
+
+    public:
+        ROCPRIM_HOST_DEVICE inline proxy_pointer(const T& value) : value_(value) {}
```
Same as the above two

amd-hsong reviewed (commented):
Looks good to me. Just wondering if it is worthwhile to pull the "proxy_pointer" class out to a common place to avoid repeating it three times. I am OK with the current way too -- it is a small piece of code anyway.

The newly added unit tests only test "->" on test_warp, it_warp etc. I wonder if the unwrapped version can be tested too.

NB4444 commented:
I have moved the proxy_pointer to a common place and added explicit, but what do you mean test the unwrapped version. The unwrapped version does not have any `value` that can be retrieved by the `->` operator.
